File: skills/faion/knowledge/dev/csharp-xunit-testing/templates/ServiceTests.cs
// purpose: Service unit test with Moq + FluentAssertions
// consumes: IOrderRepository (mocked) + IOrdersService
// produces: unit test class
// depends-on: content/01-core-rules.xml
// token-budget-impact: ~250 tokens when loaded as reference

using FluentAssertions;
using Moq;

namespace Faion.Tests;

public sealed class OrdersServiceTests
{
    private readonly Mock<IOrderRepository> _repo = new();
    private readonly OrdersService _svc;

    public OrdersServiceTests()
    {
        _svc = new OrdersService(_repo.Object);
    }

    [Fact]
    public async Task GetAsync_ExistingOrder_ReturnsDto()
    {
        _repo.Setup(r => r.GetAsync(It.Is<int>(id => id == 1), It.IsAny<CancellationToken>()))
             .ReturnsAsync(new Order(1, "Alice", 10m));

        var result = await _svc.GetAsync(1, CancellationToken.None);

        result.Should().BeEquivalentTo(new OrderResponse(1, "Alice", 10m));
        _repo.Verify(r => r.GetAsync(1, It.IsAny<CancellationToken>()), Times.Once);
    }

    [Theory]
    [InlineData(0)]
    [InlineData(-5)]
    public async Task GetAsync_NonPositiveId_Throws(int id)
    {
        var act = () => _svc.GetAsync(id, CancellationToken.None);
        await act.Should().ThrowAsync<ArgumentOutOfRangeException>();
    }
}

public interface IOrderRepository { Task<Order?> GetAsync(int id, CancellationToken ct); }
public sealed record Order(int Id, string CustomerName, decimal Total);
public sealed record OrderResponse(int Id, string CustomerName, decimal Total);
public sealed class OrdersService
{
    private readonly IOrderRepository _r;
    public OrdersService(IOrderRepository r) => _r = r;
    public async Task<OrderResponse?> GetAsync(int id, CancellationToken ct)
    {
        if (id <= 0) throw new ArgumentOutOfRangeException(nameof(id));
        var o = await _r.GetAsync(id, ct);
        return o is null ? null : new OrderResponse(o.Id, o.CustomerName, o.Total);
    }
}

File: .archive/knowledge/dev/java-junit-testing/templates/ServiceTests.java
// purpose: Mockito service unit test skeleton
// consumes: service + repository mock
// produces: service unit test class
// depends-on: content/01-core-rules.xml
// token-budget-impact: ~200 tokens when loaded as reference

package faion.app.orders;

import faion.infra.orders.OrderRepository;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Optional;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.argThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class OrderServiceTest {

    @Mock private OrderRepository repository;
    @InjectMocks private OrderService service;

    @Test
    void getById_existingOrder_returnsResponse() {
        UUID id = UUID.randomUUID();
        when(repository.findByIdWithItems(id)).thenReturn(Optional.of(new faion.domain.orders.Order("Alice")));

        OrderResponse response = service.getById(id);

        assertThat(response.customerName()).isEqualTo("Alice");
        verify(repository).findByIdWithItems(argThat(arg -> arg.equals(id)));
    }

    @Test
    void getById_missing_throws() {
        UUID id = UUID.randomUUID();
        when(repository.findByIdWithItems(id)).thenReturn(Optional.empty());

        assertThatThrownBy(() -> service.getById(id))
            .isInstanceOf(OrderNotFoundException.class);
    }
}
